tests/tests.py

```diff
@@ -938,3 +952,425 @@ def test_invalidate_model_with_geometry(self):
         geom.save()
         # Raises ValueError if this doesn't work
         invalidate_obj(geom)
+
+
+class ThreadWithReturnValue(Thread):
+    def __init__(self, *args, **kwargs):
+        super(ThreadWithReturnValue, self).__init__(*args, **kwargs)
+        self._return = None
+
+    def run(self):
+        if six.PY3:
+            if self._target is not None:
+                self._return = self._target(*self._args, **self._kwargs)
+        if six.PY2:
+            if self._Thread__target is not None:
+                self._return = self._Thread__target(*self._Thread__args, **self._Thread__kwargs)
+
+    def join(self, *args, **kwargs):
+        super(ThreadWithReturnValue, self).join(*args, **kwargs)
+        return self._return
+
+
+class _AssertNumQueriesForNonTestCaseContext(CaptureQueriesContext):
+    def __init__(self, num, using=None):
+        self.num = num
+        if not using:
+            using = DEFAULT_DB_ALIAS
+        conn = connections[using]
+        super(_AssertNumQueriesForNonTestCaseContext, self).__init__(conn)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        super(_AssertNumQueriesForNonTestCaseContext, self).__exit__(exc_type, exc_value, traceback)
+        if exc_type is not None:
+            return
+        executed = len(self)
+        if executed != self.num:
+            raise AssertionError(
+                "%d queries executed, %d expected\nCaptured queries were:\n%s" % (
+                    executed, self.num,
+                    '\n'.join(
+                        query['sql'] for query in self.captured_queries
+                    )
+                )
+            )
```

Comment:
Supported query number counting for other threads, which didn't have access to the test. Removed in b726baa1